subsys/bluetooth/host/l2cap.c

```diff
@@ -1386,36 +1430,42 @@ static void l2cap_chan_le_recv_sdu(struct bt_l2cap_le_chan *chan,
 		return;
 	}
 
-	/* Jump to last fragment */
-	frag = net_buf_frag_last(chan->_sdu);
+	/* Append received segment to SDU */
+	len = net_buf_append_bytes(chan->_sdu, buf->len, buf->data, K_NO_WAIT,
+				   l2cap_alloc_frag, chan);
+	if (len != buf->len) {
+		BT_ERR("Unable to store SDU");
+		bt_l2cap_chan_disconnect(&chan->chan);
+		return;
+	}
 
-	while (buf->len) {
-		/* Check if there is any space left in the current fragment */
-		if (!net_buf_tailroom(frag)) {
-			frag = l2cap_alloc_frag(chan);
-			if (!frag) {
-				BT_ERR("Unable to store SDU");
-				bt_l2cap_chan_disconnect(&chan->chan);
-				return;
-			}
+	if (net_buf_frags_len(chan->_sdu) < chan->_sdu_len) {
+		/* Give more credits if remote has run out of them, this
+		 * should only happen if the remote cannot fully utilize the
+		 * MPS for some reason.
+		 */
+		if (!k_sem_count_get(&chan->rx.credits)) {
+			l2cap_chan_update_credits(chan, buf);
 		}
-
-		len = min(net_buf_tailroom(frag), buf->len);
-		net_buf_add_mem(frag, buf->data, len);
-		net_buf_pull(buf, len);
-
-		BT_DBG("frag %p len %u", frag, frag->len);
+		return;
 	}
 
-	if (net_buf_frags_len(chan->_sdu) == chan->_sdu_len) {
-		/* Receiving complete SDU, notify channel and reset SDU buf */
-		chan->chan.ops->recv(&chan->chan, chan->_sdu);
-		net_buf_unref(chan->_sdu);
-		chan->_sdu = NULL;
-		chan->_sdu_len = 0;
+	/* Receiving complete SDU, notify channel and reset SDU buf */
+	err = chan->chan.ops->recv(&chan->chan, chan->_sdu);
+	net_buf_unref(chan->_sdu);
+	chan->_sdu = NULL;
+	chan->_sdu_len = 0;
+
+	if (err < 0) {
+		if (err != -EINPROGRESS) {
```

Comment:
Should you not do this for line 1513 as well?

Reply:
Nope, that is not doing segmentation in that case the channel will receive each segment and has to reassemble it.